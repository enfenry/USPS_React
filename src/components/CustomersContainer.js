"use strict"

import * as customersActions from '../actions/customersActions';
import CustomersRender from './CustomersRender';
import PropTypes from 'prop-types';
import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import LoadingIcon from './LoadingIcon';
import ErrorBanner from './ErrorBanner';

const CustomersContainer = (props) => {
    const { actions, customers, requestState } = props;
    const {
        error,

        customersCreateFailed, customersCreateSuccess,
        customersReadPending, customersReadFailed, customersReadSuccess,
        customersUpdateFailed, customersUpdateSuccess,
        customersDeleteFailed, customersDeleteSuccess,

        // Associated Entities required for displaying information in tables
        addressesReadPending, addressesReadFailed, addressesReadSuccess,
    } = requestState;

    useEffect(() => {
        actions.readCustomers();
    }, []);

    const renderSuccess = () => {
        return (
            <div className="reactive-margin">
                <CustomersRender
                    customers={customers}
                    handleUpdate={(values, customer) => {
                        actions.updateCustomer(values, customer.contactid)
                    }}
                    handleDelete={customer => {
                        actions.deleteCustomer(customer.contactid)
                    }}
                    handleCreate={(values) => {
                        actions.createCustomer(values)
                    }}
                    handleRefresh={() => props.actions.readCustomers()}
                />
            </div>
        );
    }

    if (customersReadPending || addressesReadPending) {
        return <LoadingIcon />;
    } else if (customersReadFailed || addressesReadFailed) {
        return (
            <ErrorBanner>
                Error while loading customers!
            </ErrorBanner>
        );
    } else if (customersDeleteFailed) {
        return (
            <React.Fragment>
                <ErrorBanner>
                    {error.message}
                    <br />
                    Cannot delete: Record is associated with another entity record.
                </ErrorBanner>
                {renderSuccess()}
            </React.Fragment>
        );
    } else if (customersUpdateFailed || customersCreateFailed) {
        return (
            <React.Fragment>
                <ErrorBanner>
                    {error.message}
                    <br />
                </ErrorBanner>
                {renderSuccess()}
            </React.Fragment>
        );
    } else if ((customersReadSuccess || customersCreateSuccess || customersUpdateSuccess || customersDeleteSuccess) && addressesReadSuccess) {
        return renderSuccess();
    } else {
        return (
            <ErrorBanner>
                Invalid state! This message should never appear.
            </ErrorBanner>
        );
    }
}

CustomersContainer.propTypes = {
    actions: PropTypes.object
};

function mapStateToProps(state) {
    const { customersReducer, addressesReducer } = state;
    return {
        customers: state.customersReducer.customers,
        addresses: state.addressesReducer.addresses,
        requestState: Object.assign({},
            addressesReducer.requestState,
            customersReducer.requestState)
    }
}

function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators(customersActions, dispatch)
    }
}

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(CustomersContainer);