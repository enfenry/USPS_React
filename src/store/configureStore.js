import { createStore, applyMiddleware, compose } from 'redux';
import reduxImmutableStateInvariant from 'redux-immutable-state-invariant';
import thunk from 'redux-thunk';
import rootReducer from '../reducers';

import { createLogger } from 'redux-logger'
import DevTools from '../containers/DevTools';

function configureStore(initialState) {

  const logger = createLogger()

  const middlewares = [
    reduxImmutableStateInvariant(),
    thunk,
    logger
  ];

  const store = createStore(
    rootReducer,
    initialState,
    compose(
      applyMiddleware(...middlewares),
      DevTools.instrument()
    )
  );

  return store;
}

export default configureStore;
