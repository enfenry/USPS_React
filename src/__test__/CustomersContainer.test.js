import React from 'react';
import CustomersContainer from '../components/CustomersContainer';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';

import * as basicStateUtils from "./basicState";

import configureStore from 'redux-mock-store';
const mockStore = configureStore([]);

test('CustomersContainer renders correctly', () => {

  const store = mockStore(basicStateUtils.getBasicState());
  
  const component = renderer.create(
    <Provider store={store}>
      <CustomersContainer />
    </Provider>
  );

  let tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});