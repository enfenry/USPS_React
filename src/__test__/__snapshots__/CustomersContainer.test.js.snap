// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomersContainer renders correctly 1`] = `
<div
  className="reactive-margin"
>
  <h1>
    Customers
  </h1>
  <div>
    <p>
      <button
        aria-label={null}
        className="btn btn-primary"
        onClick={[Function]}
        type="button"
      >
        Create New Customer
      </button>
    </p>
  </div>
  <div
    className="dataTables_wrapper dt-bootstrap4"
    data-test="datatable"
  >
    <div
      className="row"
    >
      <div
        className="col-sm-12 col-md-6"
        data-test="datatable-entries"
      >
        <div
          className="dataTables_length bs-select"
          data-test="datatable-select"
        >
          <label>
            Show entries
            <select
              className="custom-select custom-select-sm form-control form-control-sm"
              onChange={[Function]}
              style={
                Object {
                  "marginLeft": ".5rem",
                }
              }
              value={10}
            >
              <option
                value={10}
              >
                10
              </option>
              <option
                value={20}
              >
                20
              </option>
              <option
                value={50}
              >
                50
              </option>
              <option
                value={100}
              >
                100
              </option>
            </select>
          </label>
        </div>
      </div>
      <div
        className="col-sm-12 col-md-6"
        data-test="datatable-search"
      >
        <div
          className="dataTables_filter"
          data-test="datatable-input"
        >
          <label>
            Search
            <input
              className="form-control form-control-sm"
              onChange={[Function]}
              placeholder="Search"
              type="search"
              value=""
            />
          </label>
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-sm-12"
        data-test="datatable-table"
      >
        <div
          className="table-responsive"
          data-test="table"
          style={
            Object {
              "maxHeight": undefined,
            }
          }
        >
          <table
            className="table table-bordered table-sm table-striped dataTable"
            entries={10}
          >
            <thead
              data-test="datatable-head"
            >
              <tr>
                <th
                  className="sorting"
                  onClick={[Function]}
                >
                  ID
                </th>
                <th
                  className="sorting"
                  onClick={[Function]}
                >
                  Name
                </th>
                <th
                  className="sorting"
                  onClick={[Function]}
                >
                  
                </th>
                <th
                  className="sorting"
                  onClick={[Function]}
                >
                  
                </th>
                <th
                  className="sorting"
                  onClick={[Function]}
                >
                  
                </th>
              </tr>
            </thead>
            <tbody
              data-test="table-body"
            >
              <tr>
                <td>
                  example id
                </td>
                <td>
                  example name
                </td>
                <td>
                  <div>
                    <p>
                      <button
                        aria-label={null}
                        className="btn btn-secondary"
                        onClick={[Function]}
                        type="button"
                      >
                        View
                      </button>
                    </p>
                  </div>
                </td>
                <td>
                  <div>
                    <p>
                      <button
                        aria-label={null}
                        className="btn btn-success"
                        onClick={[Function]}
                        type="button"
                      >
                        Update
                      </button>
                    </p>
                  </div>
                </td>
                <td>
                  <div>
                    <p>
                      <button
                        aria-label={null}
                        className="btn btn-danger"
                        onClick={[Function]}
                        type="button"
                      >
                        Delete
                      </button>
                    </p>
                  </div>
                </td>
              </tr>
            </tbody>
            <thead
              data-test="table-foot"
            >
              <tr>
                <th>
                  ID
                </th>
                <th>
                  Name
                </th>
                <th>
                  
                </th>
                <th>
                  
                </th>
                <th>
                  
                </th>
              </tr>
            </thead>
          </table>
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-sm-12 col-md-5"
        data-test="datatable-info"
      >
        <div
          aria-live="polite"
          className="dataTables_info"
          role="status"
        >
          Showing 1 to 1 of 1 entries
        </div>
      </div>
      <div
        className="col-sm-12 col-md-7"
        data-test="datatable-pagination"
      >
        <div
          className="dataTables_paginate"
        >
          <ul
            className="pagination"
            data-test="pagination"
          >
            <li
              className="disabled page-item"
              data-test="page-item"
            >
              <a
                aria-label="Previous"
                className="page-link page-link"
                data-test="page-link"
                onClick={[Function]}
              >
                <span>
                  Previous
                </span>
              </a>
            </li>
            <li
              className="active page-item"
              data-test="page-item"
            >
              <a
                className="page-link page-link"
                data-test="page-link"
                onClick={[Function]}
              >
                1
                <span
                  className="sr-only"
                >
                  (current)
                </span>
              </a>
            </li>
            <li
              className="disabled page-item"
              data-test="page-item"
            >
              <a
                aria-label="Next"
                className="page-link page-link"
                data-test="page-link"
                onClick={[Function]}
              >
                <span>
                  Next
                </span>
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;
